Sources/AsyncAlgorithms/AsyncThrowingRecursiveMapSequence.swift

```diff
@@ -0,0 +1,127 @@
+    ///         View(id: 3),
+    ///         View(id: 4, children: [
+    ///             View(id: 6),
+    ///         ]),
+    ///         View(id: 5),
+    ///     ]),
+    ///     View(id: 2),
+    /// ]
+    /// for await view in tree.async.recursiveMap({ $0.children.async }) {
+    ///     print(view.id)
+    /// }
+    /// // 1
+    /// // 2
+    /// // 3
+    /// // 4
+    /// // 5
+    /// // 6
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - transform: A closure that map the element to new sequence.
+    /// - Returns: A sequence of the original sequence followed by recursive mapped sequence.
+    @inlinable
+    public func recursiveMap<C>(_ transform: @Sendable @escaping (Element) async throws -> C) -> AsyncThrowingRecursiveMapSequence<Self, C> {
+        return AsyncThrowingRecursiveMapSequence(self, transform)
+    }
+}
+
+public struct AsyncThrowingRecursiveMapSequence<Base: AsyncSequence, Transformed: AsyncSequence>: AsyncSequence where Base.Element == Transformed.Element {
+    
+    public typealias Element = Base.Element
+    
+    @usableFromInline
+    let base: Base
+    
+    @usableFromInline
+    let transform: @Sendable (Base.Element) async throws -> Transformed
+    
+    @inlinable
+    init(_ base: Base, _ transform: @Sendable @escaping (Base.Element) async throws -> Transformed) {
+        self.base = base
+        self.transform = transform
+    }
+    
+    @inlinable
+    public func makeAsyncIterator() -> AsyncIterator {
+        return AsyncIterator(base, transform)
+    }
+}
+
+extension AsyncThrowingRecursiveMapSequence {
+    
+    public struct AsyncIterator: AsyncIteratorProtocol {
+        
+        @usableFromInline
+        var base: Base.AsyncIterator?
+        
+        @usableFromInline
+        var mapped: ArraySlice<Transformed> = []
+        
+        @usableFromInline
+        var mapped_iterator: Transformed.AsyncIterator?
+        
+        @usableFromInline
+        var transform: @Sendable (Base.Element) async throws -> Transformed
+        
+        @inlinable
+        init(_ base: Base, _ transform: @Sendable @escaping (Base.Element) async throws -> Transformed) {
+            self.base = base.makeAsyncIterator()
+            self.transform = transform
+        }
+        
+        @inlinable
+        public mutating func next() async throws -> Base.Element? {
+            
+            if self.base != nil {
+                
+                if let element = try await self.base?.next() {
+                    try await mapped.append(transform(element))
+                    return element
+                }
+                
+                self.base = nil
+                self.mapped_iterator = mapped.popFirst()?.makeAsyncIterator()
+            }
+            
+            while self.mapped_iterator != nil {
+                
+                if let element = try await self.mapped_iterator?.next() {
+                    try await mapped.append(transform(element))
```

Comment:
if this throws, don't you have the iterator potentially be in a bad state? where it may throw an error for this call to next but then the next call to next violates the expectation that past a nil or a thrown error the iterators must return nil for subsequent calls to next.